### PR TITLE
Make remaining string methods work with UTF-8

### LINKED ISSUE #10479: Make remaining string methods work with UTF-8
The methods that need to be made UTF-8 aware are:
```
split(), startsWith(), endsWith(), find(), rfind()
count(), replace(), join(), strip(), partition()
```
Some of these will definitely require modification.  Some others might not.

### PR DESCRIPTION
**Pull request by dmk42**
This finishes the conversion to UTF-8 based on codepoints by making the string methods not previously converted work with UTF-8.

Testing this change found three preexisting bugs that had to be fixed.
* `wcrtomb()` was being used incorrectly.
* `qio_set_glocale()` was not always being called when needed.
* String literals were being built with malformed hex escapes.

Closes #10479

### CODE REVIEW

**Comment by dmk42** [comment]
@mppf - Would you please review?

**Comment by dmk42** [comment]
@mppf - I am having second thoughts about my implementation of the trailing part of `strip()`.  Rather than creating a new copy of the string by iterating over `localThis[start..]`, maybe it would be better to do the work on `localThis` itself and just skip the characters we find that occur before `start`.  What do you think?

File: modules/internal/String.chpl [hunk 1]
```diff
@@ -452,7 +452,16 @@ module String {
     /*
       Iterates over the string Unicode character by Unicode character.
     */
-    iter uchars(): int(32) {
+    inline iter uchars(): int(32) {
+      for (codepoint, i, nbytes) in this.ucharsIndexLen() do
+        yield codepoint;
+    }
```
Comment (mppf) [inline modules/internal/String.chpl]: does `inline iter` actually do anything?

File: modules/internal/String.chpl [hunk 2]
```diff
@@ -452,7 +452,16 @@ module String {
...
+
+    /*
+      Iterates over the string Unicode character by Unicode character,
+      and includes the byte index and byte length of each character.
+    */
+    inline iter ucharsIndexLen() {
```
Comment (mppf) [inline modules/internal/String.chpl]: I'm not sure this is a great name for it. I'd consider making this an internal function for the moment.

Comment (mppf) [reply modules/internal/String.chpl]: It also might be better to have the old iterator for performance reasons. Tuples are pretty fast but aren't free.

File: modules/internal/String.chpl [hunk 3]
```diff
@@ -936,12 +944,12 @@ module String {
                   done = true;
                 // no limit
                 } else {
-                  chunk = localThis[chunkStart..i];
+                  chunk = localThis[chunkStart..i-1];
```
Comment (mppf) [inline modules/internal/String.chpl]: Why i - 1 here? Wouldn't it need to be subtracting a codepoint size?

File: modules/internal/String.chpl [hunk 4]
```diff
@@ -1099,14 +1107,19 @@ module String {
...
-        label outer for i in 0..#localThis.len by -1 {
-          for j in 0..#localChars.len {
-            if localThis.buff[i] == localChars.buff[j] {
-              end -= 1;
+        // Because we are working with codepoints whose starting byte index
+        // is not initially known, it is faster to work forward, assuming we
+        // are already past the end of the string, and then update the end
+        // point as we are proven wrong.
+        end = 0;
+        label outer for (thisChar, i, nbytes) in localThis[start..].ucharsIndexLen() {
```
Comment (mppf) [inline modules/internal/String.chpl]: it seems to me ucharsIndexLen should be able to start at a particular byte offset (start).

File: modules/internal/String.chpl [hunk 5]
```diff
@@ -1099,14 +1107,19 @@ module String {
...
+          for removeChar in localChars.uchars() {
+            if thisChar == removeChar {
               continue outer;
             }
           }
-          break;
+          // This was not a character to be removed, so update tentative end.
+          end = start-1 + i + nbytes-1;
```
Comment (mppf) [inline modules/internal/String.chpl]: I find the two -1s here suspicious

File: modules/internal/String.chpl (hunk 1 above)
Comment (dmk42) [reply modules/internal/String.chpl]: Good point.  `linline iter` does nothing.

File: modules/internal/String.chpl (hunk 2 above)
Comment (dmk42) [reply modules/internal/String.chpl]: Thanks.  I will make ucharsIndexLen an internal function and restore the old uchars iterator.
I'm not sure there is a good name for ucharsIndexLen.

File: modules/internal/String.chpl (hunk 3 above)
Comment (dmk42) [reply modules/internal/String.chpl]: `i` was zero-based and is now one-based, so now we need `i-1`.

File: modules/internal/String.chpl (hunk 4 above)
Comment (dmk42) [reply modules/internal/String.chpl]: Yes, that will simplify this method.  Thanks.

File: modules/internal/String.chpl (hunk 5 above)
Comment (dmk42) [reply modules/internal/String.chpl]: These are three one-based numbers, so we have to subtract one from two of them to get a one-based result.

File: modules/internal/String.chpl (hunk 3 above)
Comment (dmk42) [reply modules/internal/String.chpl]: To clarify, `i` indexes the *last* byte of the *previous* character, which is why we only need to subtract 1 and not the codepoint size.

**Comment by dmk42** [comment]
@mppf - Thanks for the feedback.  This is ready for another look now.

File: modules/internal/String.chpl [hunk 6]
```diff
@@ -467,6 +467,28 @@ module String {
       }
     }
 
+    /*
+      Iterates over the string Unicode character by Unicode character,
+      and includes the byte index and byte length of each character.
+      Skip characters that begin prior to the specified start index.
```
Comment (mppf) [inline modules/internal/String.chpl]: say index here is in bytes

File: runtime/src/chpl-init.c [hunk 7]
```diff
@@ -141,6 +141,7 @@ void chpl_rt_init(int argc, char* argv[]) {
   // UTF-8 functions (e.g. wcrtomb) work as
   // indicated by the locale environment variables.
   setlocale(LC_CTYPE,"");
+  qio_set_glocale();
```
Comment (mppf) [inline runtime/src/chpl-init.c]: I think this used to be here but we moved it into qio functions e.g. qio_channel_scan_string for some reason I can't remember. Anyway I personally think it makes more sense here.

I'm fine with it being here but if it is you should remove the other calls to it.

File: test/types/string/dmk/utf8_values.chpl [hunk 8]
```diff
@@ -0,0 +1,9 @@
+var mystring1 = "événement";
+
+for c in mystring1.uchars() {
+  const str = codePointToString(c);
+  write(c, "\t", str);
+  for i in 0..#str.length do
+    write(" ", str.buff[i]:int);
+  writeln();
+}
```
Comment (mppf) [inline test/types/string/dmk/utf8_values.chpl]: You definitely need more testing - especially strip testing left and right sides. Ideally every string fn with unicode support should have a test.

**Pull request merged**